File: go-solutions/735_asteroid_collision.go
package leetcode

import "go-solutions/utils"

func AsteroidCollision(asteroids []int) []int {
	stk := utils.Stack{}
	for i := 0; i < len(asteroids); i++ {
		asteroid := asteroids[i]
		stkLen := len(stk)
		if stkLen == 0 || asteroid >= 0 || stk[stkLen-1] < 0 {
			stk.Push(asteroid)
		} else {
			if stk[len(stk)-1] < -asteroid {
				stk = stk[:len(stk)-1]
				i--
			} else if -asteroid == stk[len(stk)-1] {
				stk = stk[:len(stk)-1]
			}
		}
	}
	return stk
}

File: ts-solutions/src/735-asteroid-collision.ts
export function asteroidCollision(asteroids: number[]) {
	const result = new Array<number>()
	for (let i = 0; i < asteroids.length; i++) {
		const asteroid = asteroids[i]
		const lastAsteroid = result[result.length - 1]
		if (result.length === 0 || lastAsteroid < 0 || asteroid > 0) {
			result.push(asteroid)
		} else {
			if (-asteroid > lastAsteroid) {
				result.pop()
				i--
			} else if (-asteroid === lastAsteroid) {
				result.pop()
			}
		}
	}
	return result
}
